### PR TITLE
 feature: Use `serde_cbor` for network message encoding

### LINKED ISSUE #7: feature: Use `serde_cbor` for network message encoding.
Currently the library uses `serde_json` for encoding protocol messages. Which is useful for logging and debugging, but inefficient for network messages.

Network messages can be encoded in `serde_cbor` for more optimal encoding and logging can still use `serde_json`.

### PR DESCRIPTION
**Pull request by wthrajat**
Aims to fix #7.
Work in progress.

### CODE REVIEW

**Comment by rajarshimaitra** [comment]
Concept ACK. This will be a useful addition to have.

**Comment by wthrajat** [comment]
This became a bit worrisome to me because `CBOR` handles messages differently.

In the PR, in `src/maker/mod.rs`, on logging `buf` and then decoding it using https://cbor.me (after making it to hex)
We get 2 messages with length prefixes:
```json
{"type": "takerhello", "protocol_version_min": "protocol_version_max"}
```
and
```json
{"type": "reqgiveoffer"}
```
But `CBOR` works on one message at a time without a length prefix. Hence we needed to adjust the method to parse messages correctly.
So now the length prefix is read with
```rust
let len = u32::from_be_bytes(buf[0..4].try_into().unwrap()) as usize;
```
and then it checks if the buffer has enough data for the entire message with
```rust
if buf.len() < len + 4
```
If the buffer doesn't have enough data, it breaks the loop and waits for more data to arrive. After ensuring the buffer has enough data, it extracts the message data with
```rust
let message_data = &buf[4..4 + len];
```
and then deserializes it with
```rust
serde_cbor::from_slice(message_data);
```
And this de/serialisation works fine because the below code never gets executed and we can see the message in correct format if we log it!
```rust
 Err(e) => {
                            log::error!("Error deserializing message: {:?}", e);
                            continue;
                            }
```
Example:
```console
2024-01-09T04:18:52Z INFO  coinswap::maker] Message deserialised successfuly: TakerHello(TakerHello { protocol_version_min: 0, protocol_version_max: 0 })
[2024-01-09T04:18:52Z INFO  coinswap::maker] Message deserialised successfuly: TakerHello(TakerHello { protocol_version_min: 0, protocol_version_max: 0 })
```
And
```console
2024-01-09T04:18:52Z INFO  coinswap::maker] [16102] ===> MakerHello
[2024-01-09T04:18:52Z INFO  coinswap::utill] -----Here is message: MakerHello(
        MakerHello {
            protocol_version_min: 0,
            protocol_version_max: 0,
        },
    )
[2024-01-09T04:18:52Z INFO  coinswap::utill] -----Here is message: MakerHello(
        MakerHello {
            protocol_version_min: 0,
            protocol_version_max: 0,
        },
    )
```

What seems odd to me is that we are getting `0` offers, `untried makers count` is `0`
```console
2024-01-09T04:19:12Z INFO  coinswap::taker::api] <=== Got Offers (0 offers)
[2024-01-09T04:19:12Z INFO  coinswap::taker::api] ---Swap Locktime: 144
[2024-01-09T04:19:12Z INFO  coinswap::maker] [16102] <=== TakerHello 
[2024-01-09T04:19:12Z INFO  coinswap::maker] [6102] <=== TakerHello 
[2024-01-09T04:19:12Z INFO  coinswap::taker::api] ----Untried Makers Count: 0
[2024-01-09T04:19:12Z INFO  coinswap::taker::api] We don't have enough makers to satisfy the swap requirements!
[2024-01-09T04:19:12Z INFO  coinswap::maker] Message deserialised successfuly: TakerHello(TakerHello { protocol_version_min: 0, protocol_version_max: 0 })
[2024-01-09T04:19:12Z INFO  coinswap::maker] Message deserialised successfuly: TakerHello(TakerHello { protocol_version_min: 0, protocol_version_max: 0 })
[2024-01-09T04:19:12Z ERROR coinswap::taker::api] Could not initiate first hop: NotEnoughMakersInOfferBook
[2024-01-09T04:19:12Z INFO  coinswap::maker] [6102] Connection closed by peer
[2024-01-09T04:19:12Z INFO  coinswap::maker] [16102] Connection closed by peer
test test_stop_taker_after_setup ... FAILED
```

**Comment by rajarshimaitra** [comment]
Thanks for the detailed breakdown here. Been a bit slow on code stuffs due to other errands. Pumping this up in my todo list..

File: src/error.rs [hunk 1]
```diff
@@ -9,6 +9,7 @@ pub enum NetError {
     Json(serde_json::Error),
```
Comment (rajarshimaitra) [inline src/error.rs:9]: Do we need Json anymore?

**Review by rajarshimaitra — COMMENTED** [review]
ACK..

**Comment by rajarshimaitra** [comment]
@wthrajat updated with my code. Please check.

File: src/error.rs (hunk 1 above)
Comment (wthrajat) [reply src/error.rs]: Removed in a453f94

**Review by rajarshimaitra — APPROVED** [review]
ACK 8328094ffdf26541dea6b969f421668d102027cf

**Pull request merged**